modules/nixos/macos.nix

```diff
@@ -49,4 +50,35 @@
     programs.htop.enable = true;
     programs.htop.settings.show_program_path = true;
   };
+
+  system.activationScripts.postActivation.text = let
+    authorizedKeysDeployUser = "/etc/ssh/authorized_keys.d/${config.deployUser}";
+    sshdAuthorizedKeysConf = "/etc/ssh/sshd_config.d/200-authorized-keys.conf";
+  in ''
+    mkdir -p $(dirname ${sshdAuthorizedKeysConf})
+    echo "AuthorizedKeysFile .ssh/authorized_keys /etc/ssh/authorized_keys.d/%u" > ${sshdAuthorizedKeysConf}
+    chmod 444 ${sshdAuthorizedKeysConf}
+
+    mkdir -p $(dirname ${authorizedKeysDeployUser})
+    echo > ${authorizedKeysDeployUser}
+    chmod 444 ${authorizedKeysDeployUser}
+    ${builtins.concatStringsSep "\n" (
+      lib.attrsets.mapAttrsToList (
+        keyName: keyFile: ''
+          echo '# ${keyName}' >> ${authorizedKeysDeployUser}
+          cat ${keyFile} >> ${authorizedKeysDeployUser}
```

Comment:
Possible I'm not following this right, but how are newlines being handled here?

`echo`'ing hte keyname to the file makes sense, that will output a newline. But `cat` I wouldn't expect to unless the input file ends with a newline. Is there something going on with the `concatStringSep` I'm missing? The script lines look imperative so I'm not sure how

Reply:
here we're using nix as a templating tool to generate a shell script.
the expression `${builtins.concatStringsSep (1) (2)}` evaluates to a single string. the string is the concatenation of all elements of the list at (2) separated by (1). in our case (2) is a more complex expression but it does evaluate to a list of strings.

Reply:
here's how to see the rendered text:

```
nix-repl>                                                                                                  steveej@steveej-t14 ✓ ~/src/holo/holochain-infra pr_macos_sshd_authorized_keys ⚡
$ nix repl
Welcome to Nix 2.11.1. Type :? for help.

nix-repl> :lf .
Added 23 variables.

nix-repl> outputs.darwinConfigurations.macos-03.config.system.activationScripts.postActivation.text
"mkdir -p $(dirname /etc/ssh/sshd_config.d/200-authorized-keys.conf)\necho \"AuthorizedKeysFile .ssh/authorized_keys /etc/ssh/authorized_keys.d/%u\" > /etc/ssh/sshd_config.d/200-authorized-keys.conf\nchmod 444 /etc/ssh/sshd_config.d/200-authorized-keys.conf\n\nmkdir -p $(dirname /etc/ssh/authorized_keys.d/hetzner)\necho > /etc/ssh/authorized_keys.d/hetzner\nchmod 444 /etc/ssh/authorized_keys.d/hetzner\necho '# keys_jost-s' >> /etc/ssh/authorized_keys.d/hetzner\ncat /nix/store/h41p1vbwwx2gcq5qlkk81a2rxdfz0f00-source >> /etc/ssh/authorized_keys.d/hetzner\n\necho '# keys_maackle' >> /etc/ssh/authorized_keys.d/hetzner\ncat /nix/store/sivkdjq4b5nvf3vhi7cg8nj08p3rc0kf-source >> /etc/ssh/authorized_keys.d/hetzner\n\necho '# keys_neonphog' >> /etc/ssh/authorized_keys.d/hetzner\ncat /nix/store/1rc62xca3yqmd8pisc19qph8b2h5yykh-source >> /etc/ssh/authorized_keys.d/hetzner\n\necho '# keys_steveej' >> /etc/ssh/authorized_keys.d/hetzner\ncat /nix/store/hljnxp1rc5wkki6cpq84m2xys5bb1s3a-source >> /etc/ssh/authorized_keys.d/hetzner\n\necho '# keys_testing' >> /etc/ssh/authorized_keys.d/hetzner\ncat /nix/store/svy7w0pwpl9rx2qrwyf1q0ljki01i16f-key >> /etc/ssh/authorized_keys.d/hetzner\n\necho '# keys_thetasinner' >> /etc/ssh/authorized_keys.d/hetzner\ncat /nix/store/c3572sxfd9zs6ah03d9dhhyq05pd21kx-source >> /etc/ssh/authorized_keys.d/hetzner\n\necho '# keys_zippy' >> /etc/ssh/authorized_keys.d/hetzner\ncat /nix/store/6gg2kmrcdp6fwzq7a6nqsy8h3brgbf9p-source >> /etc/ssh/authorized_keys.d/hetzner\n\nlaunchctl kickstart -k system/com.openssh.sshd\n\n# fixup bashrc for remote nix executions\necho 'PATH=/nix/var/nix/profiles/default/bin:$PATH' > /Users/builder/.bashrc\n\n# allow builder usre to lock gc.lock\nchown builder /nix/var/nix/gc.lock\n\n# setup ssh credentials for remote builds\nmkdir -p /Users/builder/.ssh/\necho \"command=\\\"/nix/store/zdx4kqf0mkqhr8qybw2pyrdycfgsk16s-flock-0.2.3/bin/flock -s /nix/var/nix/gc.lock nix-daemon --stdio\\\" ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJ1K1ZYBnf3UqQbln5Z8DLYsXyJo6pRAFISPQ7lJZpoO root@linux-builder-01\" > /Users/builder/.ssh/authorized_keys\nchown -R builder:staff /Users/builder/.ssh/\nchmod 700 /Users/builder/.ssh/\nchmod 400 /Users/builder/.ssh/authorized_keys\ndseditgroup -o edit -a builder -t user com.apple.access_ssh\n\n# restart nix daemon if nix.conf changed\nif [ ! -e /etc/nix/nix.conf.old ] && ! diff -q /etc/nix/nix.conf /etc/nix/nix.conf.old; then\n  cp /etc/nix/nix.conf /etc/nix/nix.conf.old\n  launchctl stop org.nixos.nix-daemon\n  launchctl start org.nixos.nix-daemon\nfi\n\necho Activating home-manager configuration for hetzner\nsudo -u hetzner --set-home /nix/store/wxvyp24maqr3a492caz2ryi88nvwc6ra-activation-hetzner\n"

nix-repl>
```

notice that there is more text here because there are other places from which this attribute is set and they are joined.

Reply:
Ahh I see okay. That makes a lot more sense thank you!